### PR TITLE
PCVL-322 PPNR interleaved detectors

### PR DESCRIPTION
**Pull request by ericbrts**

### CODE REVIEW

File: tests/test_detectors.py [hunk 1]
```diff
@@ -47,6 +47,38 @@ def test_detector():
...
+    res = detector_2wires.detect(3)  # Now with 3 incoming photons
+    assert s[0] not in res
+    assert s[1] in res and res[s[1]] == pytest.approx(0.25)
+    assert s[2] in res and res[s[2]] == pytest.approx(0.75)
+    assert len(res) == 2
+
+    detector_5wires = Detector.ppnr(n_wires=5)  # Adding wires increases the probability of catching all photons
+    res = detector_5wires.detect(3)
+    assert s[0] not in res
+    assert s[1] in res and res[s[1]] == pytest.approx(0.04)
```
Comment (Aubaert) [inline tests/test_detectors.py:68]: Was this computed by hand or an independent simulation ?

Comment (ericbrts) [reply tests/test_detectors.py:68]: It was by hand, on a paper :/

File: perceval/components/detector.py [hunk 2]
```diff
@@ -132,48 +131,78 @@ def detect(self, theoretical_photons: int) -> BSDistribution | BasicState:
...
-        return BasicState([1])
+
+        if detector_type == DetectionType.Threshold:
+            return BasicState([1])
+
+        remaining_p = 1
+        result = BSDistribution()
+        max_detectable = min(self._max, theoretical_photons)
+        for i in range(1, max_detectable):
+            p_i = remaining_p * (i / self._wires) ** (theoretical_photons - i)
```
[the remark this answers is not in the document]
Comment (ericbrts) [reply perceval/components/detector.py:199]: I've added an explanatory comment

File: perceval/components/detector.py [hunk 3]
```diff
@@ -132,48 +131,78 @@ def detect(self, theoretical_photons: int) -> BSDistribution | BasicState:
...
         super().__init__()
-        assert 0 <= p_multiphoton_detection <= 1,\
-            f"A probability must be within 0 and 1 (got {p_multiphoton_detection})"
-        self._pmd = p_multiphoton_detection
-        if self.type == DetectionType.PPNR:
-            get_logger().error("Generic PPNR was not implemented yet and will behave like a threshold detector for now",
-                               channel.user)
+        assert n_wires is None or n_wires > 0, f"A detector requires at least 1 wire (got {n_wires})"
+        assert max_detections is None or n_wires is None or max_detections <= n_wires,\
+            f"Max detections has to be lower than the number of wires (got {max_detections} > {n_wires} wires)"
```
[the remark this answers is not in the document]
Comment (ericbrts) [reply perceval/components/detector.py:152]: I'm testing that "a <= b", so the error message is "that's bad a > b". I think it's consistent

File: perceval/components/detector.py [hunk 4]
```diff
@@ -192,24 +193,37 @@ def detect(self, theoretical_photons: int) -> BSDistribution | BasicState:
...
+        Hitting `i` wires with `n` photons is:
+        - hitting `i - 1` wires with `n - 1` photons AND hitting a new wire with the nth photon
+        OR
+        - hitting `i` wires with `n - 1` photons AND hitting one of the wire that were already hit with the nth photon
+        """
+        if det == 0:
+            return 1 if nph == 0 else 0
+        if nph < det:
+            return 0
+        return self._cond_probability(det - 1, nph - 1) * (self._wires - det + 1) / self._wires \
```
Comment (Aubaert) [inline perceval/components/detector.py:224]: Can we cache this?
like this, if you detect 4 photons after computing for 3 photons, you will go down to 1 photon again but all you need is the results for 3 photons.
(for example, use `lru_cache` decorator)

**Pull request merged**